### ISSUE TITLE
Tab group integration

### ISSUE THREAD

**Issue by noctuid** [cmcaine/tridactyl#811]
This may not be possible to implement currently (haven't checked if it's still possible for addons to interact with other addons), but I thought I'd at least make an issues since I didn't see any related issues. Tab groups are unfortunately no longer builtin, and the old complete replacement addon and alternatives broke with quantum. [This addon](https://addons.mozilla.org/en-US/firefox/addon/simple-tab-groups/) seems to be supported/maintained though (but with a firefox bug mentioned in the description). There is also [panorama view](https://addons.mozilla.org/en-US/firefox/addon/panorama-view/).

The basic idea is to separate tabs into named contexts. This is useful so that only tabs relevant to what you are currently doing are shown (e.g. you have a tab group for documentation/stack overflow questions about some language you're learning, another tab group related to some programming project, another for music/entertainment, etc.).

Pentadactyl has a plugin for tab group commands (which is broken), and vimperator has the functionality builtin (and working). The main useful features that could be provided are a buffer listing local to the current tab group and commands for creating and switching between tab groups. With autocommands and #808, it would also be possible to have tabgroup-local keybindings (e.g. keybindings to open commonly used urls for a project or type of activity).

Here is the documentation of vimperator's tab group commands for reference:

```
:tabgroups [sub-command]
:panorama [sub-command]
        
            Manage tab groups (also called Panorama in Firefox).
            If the [sub-command] is not specified, it lists all current tab groups.
        

        Sub commands

        add[group][!] [groupName]
                
                    Create a tab group and switches to the group.
                    If [groupName] is specified, also sets the name of the group.
                
                
                    The selected tab is moved to the group if used with [!].
                
            

        list
                
                    List current tab groups.
                
            

        pulltab[tab][!] {url|index}
                
                    Pull the tab from another group.
                    The argument {url|index} is the same as :buffer's argument.
                
                
                    Also switch to the tab if [!] is specified.
                
            

        pushtab[!] {groupName}
        stash[!] {groupName}
                
                    Move the current tab to a group named {groupName}.
                
                
                    If the group does not exist, you can use [!]
                    to automatically create a new group named {groupName} and
                    move the tab there.
                
            

        remove[group] [groupName]
        rm[group] [groupName]
                
                    Close all tabs in the group named [groupName].
                    When [groupName] is omitted, close the current group.
                
            

        rename[!] [newGroupName]
        mv[!] [newGroupName]
                
                    Rename current tab group to [newGroupName].
                
                
                    If [newGroupName] is not supplied and [!] is
                    specified, resets group's name to '(Untitled)'.
                
            

        [count]switch {groupName}
                
                    Switch to a group named {groupName}.
                
                
                    Switch to [count]th next group if [count] is specified.
```

These all have `tabgroup-<sub-command>` aliases.

**Comment by mirkojovanovic** [comment]
I really miss this.

**Comment by bovine3dom** [comment]
@mirkojovanovic this would relatively straightforward to implement with the tab hiding API.

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/hide

I'd be happy to help you if you want to have a go at implementing it yourself. I guess the first thing to try to implement would be `tabhide id` in src/excmds.ts, and you can worry about the rest later.

**Comment by mirkojovanovic** [comment]
@bovine3dom I'd love to take the challenge :smiley: Never done something similar though. I'll try to figure out how to implement `tabhide id` in `src/excmds.ts`. Thanks!

**Comment by bovine3dom** [comment]
Great! If you need any help with the basics, feel free to hop on Riot. There's usually a few of us around.

**Comment by infrascripting** [comment]
I know that [Conex](https://github.com/kesselborn/conex) is another implementation of the tab groups-like functionality.

**Comment by infrascripting** [comment]
I know that [Conex](https://github.com/kesselborn/conex) is another implementation of the tab groups-like functionality.

**Comment by noctuid** [comment]
The lack of this feature is a big painpoint for me. I tried using windows for a while but it's more convenient to be able to have dedicated keybindings to switch between specific groups. More importantly, I was having issues on an older firefox and older tridactyl where only one window would be restored when opening firefox (hopefully not an issue anymore but I haven't tried it since updating firefox). Just looked at the firefox APIs and this seems a lot simpler than I'd assumed. I've implemented some basic functions for hiding/showing tabs associated with a group name.

I thought about just having functions for showing/hiding containers, but I think more often than not, I'd want groups to be different from containers. Looking at the tabs APIs, it doesn't seem possible to associate extra information with a tab, so I've added a global Map for storing the set of tabs associated with a group. I have no idea how you would persist/restore this information. Is that something that would be easily possible (can someone point me to some resources if so)? Does anyone have a better suggestion for implementing this? Are tab ids the same after restoring tabs? Without persistence, this feature is kind of useless to me.

I was thinking about these for command names:

- tgroupcreate
- tgroupclose
- tgroupswitch
- tgrouprename - rename current group
- tgroupmove - move current tab to a different group

The way it would work would be that `:tgroupcreate` would make all tabs part of the group when creating the first group. If there is at least one group, it would open a new tab page and hide all other tabs. Tab group commands would only affect the current window. Pinned tabs would be shown regardless of group because that's how the tab hiding API works.

As for tabgroup-specific settings and keybindings, it seems like it would be sufficient to just bind a key to switch to a group and source a file with keybindings for that group. Or there could be a group switch autocommand (would work even if manually typing/selecting tabgroup). Does anyone see any possible issues with this?

**Comment by glacambre** [comment]
>Looking at the tabs APIs, it doesn't seem possible to associate extra information with a tab, so I've added a global Map for storing the set of tabs associated with a group. I have no idea how you would persist/restore this information. Is that something that would be easily possible (can someone point me to some resources if so)? Does anyone have a better suggestion for implementing this? Are tab ids the same after restoring tabs? Without persistence, this feature is kind of useless to me.

I think [sessions.setTabValue](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/sessions) persists across restarts although I'm not sure. You could also store information in browser storage with [tri.config.set](https://github.com/tridactyl/tridactyl/blob/ac33cbd88cab86988a899a760f3fc4dc8f6ac53c/src/lib/config.ts#L1232) or with the background script's [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) (although I think we use it already, so I'd recommend not touching it).

**Comment by noctuid** [comment]
Thanks, that's exactly what I wanted. I'll work on making a PR this week.

**Comment by noctuid** [comment]
I have the basic functionality working now for all commands except `:tgroupmove` but now realize there are a lot of edge cases I need to try
(e.g. the user drags a tab from another window to the current window, checking interaction with pinned tabs, etc.). Some of them may be complicated to deal with.

For example `:tab ` still shows all tabs as completions. I don't remember what pentadactyl/vimperator showed (and don't even have a firefox setup old enough to quickly test at this point). It either showed different sections or only showed the tabs in the current tab group. Even with that change, the user could still switch to a tab in another group, and I need to update things so this automatically switches tab groups.

I'm hoping all edge cases are possible to handle, but I haven't tried everything yet. One thing I'm unsure about is implementing `:tgroupmove`. Is there some way to get the last selected tab or the history of selected tabs. I could just switch to the previous tab in the group (as long as there is more than 1 tab), but I'd rather keep the same behavior that deleting a tab would have if possible (which doesn't necessarily move to the tab to the left).

I have very little typescript experience and am still trying to figure things out. What I've inferred from the existing code is that functions that run async functions without await either return the call or push the call to a list of promises and return that. This ensures that awaiting the function waits for all promises to complete before continuing. What I don't understand is `done.then(() => undefined)`. It seems like this isn't always used. Is this just to make it explicit that nothing is being returned, or what is the purpose of this?

Also, can some point me to some information on debugging? Or just getting console.log to work (are log statements removed when building?)? It's been pretty painful to have to build every time before testing things. I've ended up exporting all helper functions temporarily so I can use them in the console.

**Comment by bovine3dom** [comment]
> Is there some way to get the last selected tab or the history of selected tabs

Yes - they have a lastAccessed property. See https://github.com/tridactyl/tridactyl/blob/80113f26ddf8d9cfea45b3539d6ea0eddcd41600/src/completions/Tab.ts#L170

I don't honestly know what those done.thens are doing - I'll look into it later today. Take some comfort in that you can apparently write a lot of Tridactyl without knowing what they're doing, so you probably don't need to worry about it.

Log statements aren't removed when building. You are probably logging from the background script which is accessed with `about:debugging`, then click extensions, and inspect Tridactyl.

Exporting functions is what I do for debugging, too.

Reverting this commit improves error messages a bit: https://github.com/tridactyl/tridactyl/commit/a1648e18e8f11bf276ecc2b007324a1bdaab0757

**Comment by glacambre** [comment]
I use `done.then(() => undefined)` when multiple paths are present and 
some return a promise with something and others don't, to make sure that 
the caller won't start relying on a return value that sometimes won't be 
there.

The reason I'm not just using `return undefined` is that sometimes the 
callee is sync but it does asynchronous things that the caller might 
want to wait for.

This is pretty clear to me but I can perfectly see how it might not be 
for everyone - I guess we could start enforcing a rule like "no function 
should be sync and return a promise" and replace all `return 
done.then(() => undefined)` with `await done; return undefined`. This 
isn't exactly semantically equivalent as `await` would introduce a new 
place where context switches can happen and errors could be thrown, but 
it should be close enough that we don't notice. In general I try to 
avoid needless awaits because it's theoretically faster but it's 
probably a micro-optimization we don't actually need.

**Comment by bovine3dom** [comment]
> The reason I'm not just using `return undefined` is that sometimes the callee is sync but it does asynchronous things that the caller might want to wait for.

This sounds like a typing problem - the function should have a return type of  `| undefined`, right? It seems strange to force them to always be undefined, maybe they sometimes return useful things.

**Comment by glacambre** [comment]
> the function should have a return type of [type that might be useful] | undefined, right?

Yes, but Typescript won't prevent you from trying to access fields from a possibly undefined type.

>It seems strange to force them to always be undefined, maybe they sometimes return useful things.

I've only added these to functions which already returned undefined - the reason I did that is that I remember users reporting breakage on their `:composite tabopen | ...` commands when `:tabopen` started returning promises containing the newly-opened tab.

**Comment by noctuid** [comment]
> Log statements aren't removed when building. You are probably logging from the background script which is accessed with about:debugging, then click extensions, and inspect Tridactyl.

> Exporting functions is what I do for debugging, too.

Probably would be nice to have this in the contributing.md for noobs like me. I don't know if it's possible, but it would also be nice if there was a way to export everything for a build.

> This is pretty clear to me but I can perfectly see how it might not be
for everyone - I guess we could start enforcing a rule like "no function
should be sync and return a promise" and replace all `return
done.then(() => undefined)` with `await done; return undefined`.

It makes sense to me. I think I prefer `return done.then(() => undefined)`.


I'm trying to set things up so that everything works even when things are done outside of tridactyl.
the current tab group name is associated with every new tab (even tabs opened with the firefox GUI). I am using an event listener with `DOMContentLoaded` right now to associate the current tab group name with every new tab (even if the tab is opened from the firefox GUI, for example).  I'm thinking I can use `visibilitychange` to detect that someone switched to a tab in another group with `:tab`. Any ideas on the right event to detect someone moved a tab from another window into the current window (it doesn't need to detect only that case)?

Also is there some way I can make a helper function usable in both background and content?

**Comment by glacambre** [comment]
>I am using an event listener with DOMContentLoaded right now to associate the current tab group name with every new tab

I might not have an idea precise enough of what you're trying to achieve so ignore me if I'm wrong, but it sounds like [browser.tabs.onCreated](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/onCreated) would be a better fit for this.

> I'm thinking I can use visibilitychange to detect that someone switched to a tab in another group

You could also use [browser.tabs.onActivated](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/onAttached).

>Any ideas on the right event to detect someone moved a tab from another window into the current window (it doesn't need to detect only that case)?

[browser.tabs.onDetached](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/onDetached) might be a good fit for this.

>Also is there some way I can make a helper function usable in both background and content?

You can either put it in `src/lib/` and import it in whatever file you need, or put it in `excmds.ts` and add a `//#both` comment I think :).

**Comment by noctuid** [comment]
Thanks, I think those are what I need. Is there something I can hook into on firefox startup so I can show only the last active tab group?

> You can either put it in src/lib/ and import it in whatever file you need, or put it in excmds.ts and add a //#both comment I think :).

Do I need to use `browserBg` instead of `browser` to be able to use the functions in content?

**Comment by noctuid** [comment]
Also, I was having trouble getting the mode indicator to show the current tab group. Right now I'm storing the tab group with setWindowValue so it persists. If I make the mode indicator function async and read this value, I can show it in the mode indicator, but it is not updated correctly. Any suggestions?

**Comment by glacambre** [comment]
>Is there something I can hook into on firefox startup so I can show only the last active tab group?

Just putting your code somewhere in background.ts should be enough.

>Do I need to use browserBg instead of browser to be able to use the functions in content?

Yes.

>Any suggestions?

I'm not sure I see what you mean, could you open a PR?

**Comment by bovine3dom** [comment]
> Probably would be nice to have this in the contributing.md for noobs like me.

It would be really helpful if you could make a PR that puts all the stuff you wished you had known into that file. We've been developing Tridactyl for too long and so all the surprising things aren't surpising any more : )

**Comment by noctuid** [comment]
> I'm not sure I see what you mean, could you open a PR?

It's not done, but I think it's ready for some review now.

> It would be really helpful if you could make a PR that puts all the stuff you wished you had known into that file. We've been developing Tridactyl for too long and so all the surprising things aren't surpising any more : )

I don't think I understand everything I've found surprising well enough to write documentation for it, but I can at least let you know about those things.

**Comment by noctuid** [comment]
Awesome! Great to see this merged.
